Library/Homebrew/formulary.rb

```diff
@@ -283,6 +283,8 @@ def formula_name_path(tapped_name, warn: true)
     end
 
     def get_formula(spec, alias_path: nil, force_bottle: false, flags: [])
+      tap.install unless tap.installed?
```

Comment:
Seems a bit strange for this to hit the network without any notification/configuration. Could there be a separate method that could be called by e.g. `to_formulae_and_casks` instead e.g. `tap_if_needed!`?

Reply:
Well, there is already an automatic network call if a formula was migrated to another tap. Also, `CaskLoader` has done this for several years now without any problem.

Reply:
I think it'd be nice to have it be separate call for this new case. The tap migration case is less common (and, ideally, that'd also be a separate call).

Reply:
> I think it'd be nice to have it be separate call for this new case.

I don't see a reason why a non-auto-tapping version needs to exist. In that case it would just fail like it does currently.

Reply:
I don't think all uses of this (public) API want to have this auto-tapping behaviour. For one example: during formula installation referencing another formula in another tap should not tap it during the installation process.

Reply:
I don't see the problem there. In most cases at that point the tap is already installed anyways and this will be a no-op.

Reply:
Yes and not in all cases hence not wanting to make an unexpected change to a public API.

Reply:
I still don't see the problem. Worst case is that a tap is installed during installation, rather than failing the build?

Reply:
I consider the `CaskLoader` behaviour to be a bad API. I consider the current `Formulary` API not tapping to be a better API. I consider changing the behaviour of the API in this way to be a breaking (at worst) or unexpected (at best) change. You and we don't agree: this is clear.

Please either make a change similar to the one proposed above or revert f54b458cdaec5091cf7abfa3e09e9050a94a4a26 until we can find a solution that we're both happy with.

Reply:
> I consider the `CaskLoader` behaviour to be a bad API.

I don't, so I think we should let @Homebrew/tsc decide this.

I'll submit a separate PR using the previous logic in the meantime.

Reply:
@reitermarkus it's not clear to me what the question is.
The question is whether the install method should automatically tap taps for dependencies in untapped taps? Can you give a summary of what you want to achieve (behaviour before, behaviour after)

Reply:
`Formulary` should automatically install a tap (if it isn't already) when loading up a formula with the fully qualified name. `CaskLoader` does the same (and has done for years).

Reply:
Thanks. Then I don't think I agree that this behaviour is desirable: the user has not necessarily viewed the formula content, and has not agreed to external taps being installed. I don't think we want to do that behind their back.

Whether it's a good thing that `CaskLoader` does it, I can't tell, I'm rarely using casks…

Reply:
> the user has not necessarily viewed the formula content

That's already the case with how it works now. `brew install my/tap/formula` will install the tap in any case.

Reply:
With `brew install my/tap/formula` the user knows that they're install something from `my/tap`. If I type `brew install markus/tap/formula` and it installs a `notmarkus/tap` I would find that worrying as a user.

Reply:
I'm in agreement. It would be surprising for an unrelated tap to be tapped in this case, and I'd rather we not go in that direction.

Reply:
The best UI I can think of is an error message that clearly tells you the formula depends on a formula in another tap, and you need to load it. Not sure if that's what we currently have, but I would think this is the best compromise: give the user a clear indication of what to do.

Reply:
> If I type `brew install markus/tap/formula` and it installs a `notmarkus/tap` I would find that worrying as a user.

This is already the case if `markus/tap/formula` depends on `notmarkus/tap/formula`, so still not a new issue.

Reply:
If it has a dependency, yes. If it has a reference to `Formula["notmarkus/tap/formula"`: no.

I'd also be personally in favour of not auto-tapping dependencies, either.

Reply:
> The best UI I can think of is an error message that clearly tells you the formula depends on a formula in another tap, and you need to load it.

The existing message is quite clear about that:

```
TapFormulaUnavailableError (No available formula with the name "mistydemeo/blah/blah" )
Please tap it and then try again: brew tap mistydemeo/blah
```

Reply:
> I don't, so I think we should let @Homebrew/tsc decide this.

@reitermarkus We seem to have the majority of the TSC in agreement that this should not be the approach taken here. Could you adjust this PR accordingly when you get the chance? Thanks!

Reply:
@reitermarkus Bump on this. Please either do this or close out this PR and let me know and I'll do it, thanks!